[RFC] Add caching heuristic for skipping fast graphs

oulgen opened the pull request:
Summary:
Looking at scuba dataset, we can see that many graphs take less than 1s to compile. Caching is not worth it for those graphs. Lets add a heuristic.

https://fburl.com/scuba/dynamo_compile/hbf39dli

Test Plan: adhoc testing

Differential Revision: D64995390




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @aakhundov

masnesral commented:
Seems reasonable. Could use a small unit test for this path?

eellison commented on torch/_inductor/codecache.py:
```diff
@@ -1209,6 +1209,23 @@ def post_compile(
+    @staticmethod
+    def __should_save_graph(compiled_graph: CompiledFxGraph):
+        # Do not use any heuristics in unit tests
+        if torch._utils_internal.is_fb_unit_test():
```
see also `IS_CI`

oulgen replied:
is that something ok to read in codecache? It looks like something only meant to be used in unit tests

oulgen replied:
Since IS_IC is OSS only, we dont need to add it since this was to guard against JK changes.

eellison replied:
Really would prefer not adding skew between internal tests and oss tests - that leads to all sorts of problems and is hard to debug

masnesral replied:
OSS unit tests will currently save because `minimum_time=None` and `justknobs_getval_int` will return 0 in OSS? Maybe future proof if we change the default to something other than None by returning True here if `is_fb_unit_test()` or is OSS test? I think that's what Elias is saying?

c00w replied:
Do the OSS tests have a backend to save to? (I'm surprised replacing is_fbcode with IS_CI has any impact on OSS tests).

oulgen replied:
we have a redis implementation we generally mock out but there's also a local cache

oulgen commented:
@masnesral unit test and scuba logging added!

ezyang commented on torch/_utils_internal.py:
```diff
@@ -140,7 +140,7 @@ def log_trace_structured_event(*args, **kwargs) -> None:
 
 
-def log_cache_bypass(*args, **kwargs) -> None:
+def log_cache_operation(*args, **kwargs) -> None:
```
Can we have a more detailed signature here? Makes it easier to typecheck sites from oss only

ezyang commented on test/inductor/test_codecache.py:
```diff
@@ -442,6 +442,30 @@ def fn(x, y):
+    @config.patch({"fx_graph_cache": True})
+    @config.patch({"fx_graph_remote_cache": False})
+    @config.patch(
+        {"fx_graph_cache_minimum_inductor_compile_time_for_caching_ms": 100000}
```
nit: fyi you can patch these all in one go by putting them in the same dict

ezyang commented:
I guess I'm just not that enthusiastic about the version of this that (1) depends on the actual time and (2) only saves you a write on the fast graphs, and not the read as well. It adds extra operational complexity on reasoning about why something didn't upload to cache in a nondeterministic manner.  Can we characterize how much this is actually buying us?

oulgen committed:
[RFC] Add caching heuristic for skipping fast graphs (#138938)

Summary:

Looking at scuba dataset, we can see that many graphs take less than 1s to compile. Caching is not worth it for those graphs. Lets add a heuristic.

https://fburl.com/scuba/dynamo_compile/hbf39dli

Test Plan: adhoc testing

Reviewed By: masnesral

Differential Revision: D64995390

oulgen commented:
@ezyang 

I have looked into using various other heuristics so we can skip the cache all together on fast graphs but this problem essentially boils down to not being able to predict the future. Using time is one accurate way of determining this, while I agree that this makes us do extra work, such as the first cache miss. I have also looked at approximating using other data but all my attempts resulted in expensive graphs being accidentally skipped, so I think this current solution is the best one we have.

According to james march's data, there's a ton of these "fast" graphs that we are not saving anything but still paying overhead: https://fburl.com/scuba/dynamo_compile/bt6vhfj7

In terms of complexity, I'm logging these to scuba, so hopefully that will alleviate it somewhat.

c00w commented:
@oulgen  - Is there a proxy for graph size? I assume that for small graphs, a manifold network read is going be in the 10s of milliseconds at least, even if we miss, and if we could find something to skip graph reads, even if it's not perfect, it might help with skipping reads.

oulgen commented:
> @oulgen - Is there a proxy for graph size? I assume that for small graphs, a manifold network read is going be in the 10s of milliseconds at least, even if we miss, and if we could find something to skip graph reads, even if it's not perfect, it might help with skipping reads.

1 node graph can take over few minutes to compile unfortunately depending on what that node is, e.g. custom op or HOP, so there's no easy proxy. We can write some complicated heuristic to catch a lot of cases but it will be really hard to tell what we missed or caught.

@ezyang and I chatted offline to land this and see whether remote cache put time SUM reduces in aggregate. I'll land this once we are fully on zippy since we dont have experiments to conflate
